### PR TITLE
Enable clang-tidy check in CI

### PR DESCRIPTION
**Pull request by rdspring1**
- Checks only `csrc` folder
- Excludes `benchmark` and `test` folders
- Only modified files between given branch and the `main` branch are checked.
- The clang-tidy CI runtime is proportional to the number of modified files. About 20 minutes on average. 1 hour for all files in `csrc`.

### CODE REVIEW

File: csrc/contiguity.cpp [hunk 1]
```diff
@@ -527,7 +527,7 @@ void ContigIDs::build(const std::vector<IterDomain*>& ids) {
     // rfactor root domains, which should just return "zero"
     // RootAxisInfo. This should be safe as no rfactor tensor should
     // need halo.
-    if (*root_contiguity_.at(root_domain_i) &&
+    if (root_contiguity_.at(root_domain_i).value_or(false) &&
```
Comment (zasdfgbnm) [inline csrc/contiguity.cpp:530]: I think you should `TORCH_INTERNAL_ASSERT(root_contiguity_.at(root_domain_i).has_value())`. Having no value mean doesn't mean it is discontiguous, but instead, it means a bug.

File: csrc/contiguity.cpp [hunk 2]
```diff
@@ -617,7 +617,7 @@ void ContigIDs::handle(Merge* merge) {
       // If we're computing predicates (ignore_consistent_ordering_==true),
       // then we don't have this same constraint, we can just ignore
       // contiguity of the roots all together.
-      if (!*root_contiguity_.at(root_id_i) && is_indexing_pass) {
+      if (!root_contiguity_.at(root_id_i).value_or(false) && is_indexing_pass) {
```
Comment (zasdfgbnm) [inline csrc/contiguity.cpp:620]: Same here

**Review by zasdfgbnm — CHANGES_REQUESTED** [review]

File: csrc/contiguity.cpp (hunk 1 above)
Comment (jjsjann123) [reply csrc/contiguity.cpp:530]: explicit assert sounds good~ The bar is that we don't want to change behavior just for format changes. is calling `value()` directly triggering any clang-tidy error?

Comment (rdspring1) [reply csrc/contiguity.cpp]: The clang-tidy error is:
```
  Error (CLANGTIDY) [bugprone-unchecked-optional-access,-warnings-as-errors]
    unchecked access to optional value
```

We need to acquire the optional value before checking it.

File: csrc/index_compute.cpp [hunk 3]
```diff
@@ -1421,7 +1421,7 @@ std::vector<Val*> Index::getGlobalProducerStridedIndices(
       strides[dim] = cur_contig_stride->fusion()->zeroVal();
       TORCH_INTERNAL_ASSERT(
           !producer_tv->domain()->contiguity().at(dim).has_value());
-    } else if (*producer_tv->domain()->contiguity().at(dim)) {
+    } else if (producer_tv->domain()->contiguity().at(dim).value_or(false)) {
```
Comment (zasdfgbnm) [inline csrc/index_compute.cpp:1424]: I think you need to change all the contiguity handling.

**Review by zasdfgbnm — APPROVED** [review]

**Pull request merged**